### ISSUE TITLE
May bug in evhttp_connection_free_on_completion when the evhttp_connection fail to connect

### ISSUE THREAD

**Issue by lymslive** [libevent/libevent#1535]
Scenario like this:

0. I use libevent(version 2.1.7.0.1) in a C++ project, and write a method referred to the `sample/https-client.c`
1. Has called `evhttp_connection_free_on_completion` along with `evhttp_make_request` instead of explicit `evhttp_connection_free`.
2. Use valgrind to test some of my unit-test cases and found that when I give an unreachable url it reports memory leak, while in normal case when the request finished successfully, there is no memory leak. 
3. In the failure case, there is log reported something like: http socket error = Transport endpoint is not connected (107)

I've noticed the documentation on `evhttp_connection_free_on_completion` that says the evhttp connection object will be freed after the last request has completed or failed. I doubt that if the connection has not established at all, the auto free mechanism seems not take effect.

**Comment by lymslive** [comment]
Additional updates:

Actually, I used the last stable version `libevent-2.1.12`  (libevent-2.1.so.7.0.1).
When I pull the latest master branch, and test the `sample/https-client.c` with `evhttp_connection_free_on_completion` replacing of `evhttp_connection_free`, just find the memory leak no longer reproduced. So the problem should be solved.

```diff
diff --git a/sample/https-client.c b/sample/https-client.c
index 339f1de0..e86cad14 100644
--- a/sample/https-client.c
+++ b/sample/https-client.c
@@ -617,6 +617,7 @@ main(int argc, char **argv)
                fclose(f);
        }

+       evhttp_connection_free_on_completion(evcon);
        r = evhttp_make_request(evcon, req, data_file ? EVHTTP_REQ_POST : EVHTTP_REQ_GET, uri);
        if (r != 0) {
                fprintf(stderr, "evhttp_make_request() failed\n");
@@ -629,8 +630,6 @@ main(int argc, char **argv)
 error:
        ret = 1;
 cleanup:
-       if (evcon)
-               evhttp_connection_free(evcon);
        if (http_uri)
                evhttp_uri_free(http_uri);
        if (base)
```

**Comment by azat** [comment]
Great, thank you for testing, the fix is eee26deed38fc7a6b6780b54628b007a2810efcd
